File: cpp/Isomorphic_Strings.cpp
/*

You are here!
Your runtime beats 64.89 % of cpp submissions.
You are here!
Your memory usage beats 88.81 % of cpp submissions.
*/

class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.size() != t.size()) {
            return false;
        }
        unordered_map<char, char> s2t;
        unordered_map<char, char> t2s;
        for (int i = 0; i < s.size(); i++) {
            if (s2t.count(s[i]) == 0) {
                if (t2s.count(t[i]) != 0) {
                    return false;
                }
                s2t[s[i]] = t[i];
                t2s[t[i]] = s[i];                
            } else {
                if (s2t[s[i]] != t[i]) {
                    return false;
                }
            }   
        }
        return true;
    }
};



/*
You are here!
Your runtime beats 97.00 % of cpp submissions.
You are here!
Your memory usage beats 88.81 % of cpp submissions.
*/
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.length() != t.length()) return false;
        vector<int> m1(256, -1);
        vector<int> m2(256, -1);
        for (int i = 0; i < s.length(); i++) {
            if (m1[s[i]] != m2[t[i]]) return false;
            m1[s[i]] = i;
            m2[t[i]] = i;
        }
        return true;
    }
};

File: explore/july/Isomorphic_Strings.py
'''
Hash Table
Time: O(N)
Space: O(N)

You are here!
Your runtime beats 88.16 % of python3 submissions.
You are here!
Your memory usage beats 47.08 % of python3 submissions.
'''
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        s2t, t2s = {}, {}
        for ss, tt in zip(s, t):
            if ss not in s2t and tt not in t2s:
                s2t[ss] = tt
                t2s[tt] = ss
            elif ss in s2t and tt in t2s:
                if s2t[ss] == tt and t2s[tt] == ss:
                    continue
                else:
                    return False
            else:
                return False
            
        return True
        